controllers/appmesh/virtualnode_controller.go

```diff
@@ -18,36 +18,104 @@ package controllers
+	return &virtualNodeReconciler{
+		k8sClient:                    k8sClient,
+		vnResManager:                 vnResManager,
+		enqueueRequestsForMeshEvents: virtualnode.NewEnqueueRequestsForMeshEvents(k8sClient, log),
+		log:                          log,
+	}
 }
 
-// +kubebuilder:rbac:groups=appmesh.k8s.aws,resources=virtualnodes,verbs=get;list;watch;create;update;patch;delete
-// +kubebuilder:rbac:groups=appmesh.k8s.aws,resources=virtualnodes/status,verbs=get;update;patch
+// virtualNodeReconciler reconciles a VirtualNode object
+type virtualNodeReconciler struct {
+	k8sClient    client.Client
+	vnResManager virtualnode.ResourceManager
 
-func (r *VirtualNodeReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("virtualnode", req.NamespacedName)
+	enqueueRequestsForMeshEvents handler.EventHandler
+	log                          logr.Logger
+}
 
-	// your logic here
+// +kubebuilder:rbac:groups=appmesh.k8s.aws,resources=virtualnodes,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=appmesh.k8s.aws,resources=virtualnodes/status,verbs=get;update;patch
 
-	return ctrl.Result{}, nil
+func (r *virtualNodeReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	return runtime.HandleReconcileError(r.reconcile(req), r.log)
 }
 
-func (r *VirtualNodeReconciler) SetupWithManager(mgr ctrl.Manager) error {
+func (r *virtualNodeReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
-		For(&appmeshv1beta2.VirtualNode{}).
+		For(&appmesh.VirtualNode{}).
+		Watches(&source.Kind{Type: &appmesh.Mesh{}}, r.enqueueRequestsForMeshEvents).
 		Complete(r)
 }
+
+func (r *virtualNodeReconciler) reconcile(req ctrl.Request) error {
+	ctx := context.Background()
+	vn := &appmesh.VirtualNode{}
+	if err := r.k8sClient.Get(ctx, req.NamespacedName, vn); err != nil {
+		return client.IgnoreNotFound(err)
+	}
+	if !vn.DeletionTimestamp.IsZero() {
+		return r.cleanupVirtualNode(ctx, vn)
+	}
+	return r.reconcileVirtualNode(ctx, vn)
+}
+
+func (r *virtualNodeReconciler) reconcileVirtualNode(ctx context.Context, vn *appmesh.VirtualNode) error {
+	if err := r.addFinalizers(ctx, vn); err != nil {
+		return err
+	}
+	if err := r.vnResManager.Reconcile(ctx, vn); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *virtualNodeReconciler) cleanupVirtualNode(ctx context.Context, vn *appmesh.VirtualNode) error {
+	if err := r.vnResManager.Cleanup(ctx, vn); err != nil {
+		return err
+	}
+	if err := r.removeFinalizers(ctx, vn); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *virtualNodeReconciler) addFinalizers(ctx context.Context, vn *appmesh.VirtualNode) error {
+	if k8s.HasFinalizer(vn, k8s.FinalizerAWSResources) {
+		return nil
+	}
+	oldVN := vn.DeepCopy()
+	controllerutil.AddFinalizer(vn, k8s.FinalizerAWSResources)
+	if err := r.k8sClient.Patch(ctx, vn, client.MergeFrom(oldVN)); err != nil {
```

Comment:
Why is the `client.MergeFrom()` needed both here and when deleting finalizers?

Reply:
Ok, discussed offline. It's an optimization to generate and apply the diff between the oldVN and the change that will be applied by adding or deleting the Finalizer, not reapply the whole config.